File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/selection/validate_d3_d5_discriminator.py
from __future__ import annotations



from typing import Any



from .d3_d5_discriminator_fingerprint_v1 import rule4_d3d5_discriminator_fingerprint_v1_hash

from .d3_d5_evidence_provenance import (

    envelope_d08_matches_authoritative,

    envelope_evidence_provenance_matches_authoritative,

    validate_d3_d5_evidence_provenance,

)

from .q7bf_gate_ids import RULE4_Q7BF_MANDATORY_GATE_IDS, RULE4_Q7BF_MANDATORY_GATE_ID_SET





def _q7bf_ledger_pass(gates: list[dict]) -> dict[str, Any]:

    by_id = {g["gate_id"]: g for g in gates}

    for gate_id in RULE4_Q7BF_MANDATORY_GATE_IDS:

        gate = by_id.get(gate_id)

        if not gate:

            return {"ok": False, "code": "Q7BF_GATE_MISSING"}

        if gate.get("outcome") != "PASS":

            return {"ok": False, "code": "Q7BF_GATE_NOT_PASS"}

    for gate in gates:

        if gate["gate_id"] not in RULE4_Q7BF_MANDATORY_GATE_ID_SET:

            return {"ok": False, "code": "Q7BF_GATE_UNKNOWN"}

    return {"ok": True}





def _sensitivity_close_aligned(envelope: dict) -> bool:

    sens = envelope["sensitivity_assessment_status"]

    close = envelope["close_d05_discriminator_status"]

    if sens == "CONTRADICTORY" or close == "CONTRADICTORY":

        return False

    if sens == "ASSESSED_D5_QUALIFIED":

        return close == "QUALIFIES_D5"

    if sens == "ASSESSED_D5_NOT_QUALIFIED":

        return close == "QUALIFIES_D3"

    return False





def reject_legacy_d3_d5_selection_record(legacy: dict | None) -> str | None:

    if not legacy:

        return None

    return "D3_D5_LEGACY_BOOLEAN_AUTHORITY_REJECTED"





def validate_d3_d5_discriminator_envelope(

    envelope: dict | None,

    ctx: dict,

) -> dict[str, Any]:

    if not envelope:

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": ["D3_D5_DISCRIMINATOR_ENVELOPE_MISSING"],

            "discriminator_fingerprint": None,

        }



    if (

        envelope.get("formula_slot_id") != ctx["formula_slot_id"]

        or envelope.get("formula_target_id") != ctx["formula_target_id"]

    ):

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": ["D3_D5_ENVELOPE_BINDING_INVALID"],

            "discriminator_fingerprint": None,

        }



    if envelope.get("upstream_phase7_eligibility_fingerprint") != ctx.get(

        "upstream_phase7_eligibility_fingerprint"

    ):

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": ["D3_D5_ENVELOPE_BINDING_INVALID"],

            "discriminator_fingerprint": None,

        }



    if envelope.get("binding_status") != "BOUND":

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": ["D3_D5_ENVELOPE_BINDING_INVALID"],

            "discriminator_fingerprint": None,

        }



    q7 = _q7bf_ledger_pass(list(envelope.get("q7bf_gate_results") or []))

    if not q7["ok"]:

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": [q7["code"]],

            "discriminator_fingerprint": None,

        }



    provenance = validate_d3_d5_evidence_provenance(envelope, ctx)

    if not provenance["ok"]:

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": [provenance["reason_code"]],

            "discriminator_fingerprint": None,

        }



    if not (envelope.get("evidence_provenance") or []):

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": ["D3_D5_EVIDENCE_PROVENANCE_MISSING"],

            "discriminator_fingerprint": None,

        }



    d08_auth = envelope_d08_matches_authoritative(envelope, provenance)

    if not d08_auth["ok"]:

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": [d08_auth["reason_code"]],

            "discriminator_fingerprint": None,

        }



    prov_bind = envelope_evidence_provenance_matches_authoritative(

        envelope, provenance["records"]

    )

    if not prov_bind["ok"]:

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": [prov_bind["reason_code"]],

            "discriminator_fingerprint": None,

        }



    expected_fp = rule4_d3d5_discriminator_fingerprint_v1_hash(envelope)

    if envelope.get("discriminator_fingerprint") != expected_fp:

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": ["D3_D5_DISCRIMINATOR_FINGERPRINT_INVALID"],

            "discriminator_fingerprint": None,

        }



    sens = envelope["sensitivity_assessment_status"]

    if sens in {"MISSING", "NOT_EVALUATED", "INVALID", "AMBIGUOUS", "CONTRADICTORY"}:

        code = (

            "D3_D5_SENSITIVITY_MISSING"

            if sens in {"MISSING", "NOT_EVALUATED"}

            else "D3_D5_SENSITIVITY_AMBIGUOUS"

        )

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": [code],

            "discriminator_fingerprint": expected_fp,

        }



    if (

        envelope["close_d05_discriminator_status"] in {"CONTRADICTORY", "UNRESOLVED"}

        or not _sensitivity_close_aligned(envelope)

    ):

        return {

            "status": "unresolved",

            "dilution": None,

            "reason_codes": ["D3_D5_DISCRIMINATOR_CONTRADICTORY"],

            "discriminator_fingerprint": expected_fp,

        }



    if (

        sens == "ASSESSED_D5_QUALIFIED"

        and envelope["close_d05_discriminator_status"] == "QUALIFIES_D5"

    ):

        return {

            "status": "resolved",

            "dilution": "D5",

            "reason_codes": ["CLOSE_D05_QUALIFIES_D5"],

            "discriminator_fingerprint": expected_fp,

        }



    if (

        sens == "ASSESSED_D5_NOT_QUALIFIED"

        and envelope["close_d05_discriminator_status"] == "QUALIFIES_D3"

    ):

        return {

            "status": "resolved",

            "dilution": "D3",

            "reason_codes": ["CLOSE_D05_D3_PATH"],

            "discriminator_fingerprint": expected_fp,

        }



    return {

        "status": "unresolved",

        "dilution": None,

        "reason_codes": ["D3_D5_DISCRIMINATOR_UNRESOLVED"],

        "discriminator_fingerprint": expected_fp,

    }

File: packages/clinical-contracts/src/rule4/selection/validateD3D5Discriminator.ts
import type { Rule4EvidenceAdapterOutput } from '../evidence/types.js';
import type { Rule4GateResult } from '../eligibility/types.js';
import { rule4D3D5DiscriminatorFingerprintV1Hash } from './d3D5DiscriminatorFingerprintV1.js';
import {
  envelopeD08MatchesAuthoritative,
  envelopeEvidenceProvenanceMatchesAuthoritative,
  validateD3D5EvidenceProvenance,
} from './d3D5EvidenceProvenance.js';
import { RULE4_Q7BF_MANDATORY_GATE_ID_SET, RULE4_Q7BF_MANDATORY_GATE_IDS } from './q7bfGateIds.js';
import type {
  Rule4D3D5DiscriminatorEnvelope,
  Rule4D3D5SelectionRecord,
  Rule4FrozenDilution,
} from './types.js';

export type Rule4D3D5DiscriminatorValidationContext = {
  formulaSlotId: string;
  formulaTargetId: string;
  upstreamPhase7EligibilityFingerprint: string;
  evidenceAdapter?: Rule4EvidenceAdapterOutput | null;
  evidenceItems?: readonly import('../evidence/types.js').Rule4EvidenceItemEnvelope[] | null;
};

function q7bfLedgerPass(gates: readonly Rule4GateResult[]): { ok: boolean; code?: string } {
  const byId = new Map(gates.map((g) => [g.gateId, g]));
  for (const gateId of RULE4_Q7BF_MANDATORY_GATE_IDS) {
    const g = byId.get(gateId);
    if (!g) {
      return { ok: false, code: 'Q7BF_GATE_MISSING' };
    }
    if (g.outcome !== 'PASS') {
      return { ok: false, code: 'Q7BF_GATE_NOT_PASS' };
    }
  }
  for (const g of gates) {
    if (!RULE4_Q7BF_MANDATORY_GATE_ID_SET.has(g.gateId)) {
      return { ok: false, code: 'Q7BF_GATE_UNKNOWN' };
    }
  }
  return { ok: true };
}

function sensitivityCloseAligned(envelope: Rule4D3D5DiscriminatorEnvelope): boolean {
  const s = envelope.sensitivityAssessmentStatus;
  const c = envelope.closeD05DiscriminatorStatus;
  if (s === 'CONTRADICTORY' || c === 'CONTRADICTORY') {
    return false;
  }
  if (s === 'ASSESSED_D5_QUALIFIED') {
    return c === 'QUALIFIES_D5';
  }
  if (s === 'ASSESSED_D5_NOT_QUALIFIED') {
    return c === 'QUALIFIES_D3';
  }
  return false;
}

export function rejectLegacyD3D5SelectionRecord(
  legacy: Rule4D3D5SelectionRecord | null | undefined,
): string | null {
  if (!legacy) {
    return null;
  }
  return 'D3_D5_LEGACY_BOOLEAN_AUTHORITY_REJECTED';
}

export function validateD3D5DiscriminatorEnvelope(
  envelope: Rule4D3D5DiscriminatorEnvelope | null | undefined,
  ctx: Rule4D3D5DiscriminatorValidationContext,
): {
  status: 'resolved' | 'unresolved';
  dilution: Rule4FrozenDilution | null;
  reasonCodes: string[];
  discriminatorFingerprint: string | null;
} {
  const reasons: string[] = [];
  if (!envelope) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: ['D3_D5_DISCRIMINATOR_ENVELOPE_MISSING'],
      discriminatorFingerprint: null,
    };
  }

  if (
    envelope.formulaSlotId !== ctx.formulaSlotId ||
    envelope.formulaTargetId !== ctx.formulaTargetId
  ) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: ['D3_D5_ENVELOPE_BINDING_INVALID'],
      discriminatorFingerprint: null,
    };
  }

  if (envelope.upstreamPhase7EligibilityFingerprint !== ctx.upstreamPhase7EligibilityFingerprint) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: ['D3_D5_ENVELOPE_BINDING_INVALID'],
      discriminatorFingerprint: null,
    };
  }

  if (envelope.bindingStatus !== 'BOUND') {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: ['D3_D5_ENVELOPE_BINDING_INVALID'],
      discriminatorFingerprint: null,
    };
  }

  const q7 = q7bfLedgerPass(envelope.q7bfGateResults);
  if (!q7.ok) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: [q7.code!],
      discriminatorFingerprint: null,
    };
  }

  const provenance = validateD3D5EvidenceProvenance(envelope, ctx);
  if (!provenance.ok) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: [provenance.reasonCode!],
      discriminatorFingerprint: null,
    };
  }

  if (!envelope.evidenceProvenance?.length) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: ['D3_D5_EVIDENCE_PROVENANCE_MISSING'],
      discriminatorFingerprint: null,
    };
  }

  const d08Auth = envelopeD08MatchesAuthoritative(envelope, provenance);
  if (!d08Auth.ok) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: [d08Auth.reasonCode!],
      discriminatorFingerprint: null,
    };
  }

  const provBind = envelopeEvidenceProvenanceMatchesAuthoritative(envelope, provenance.records);
  if (!provBind.ok) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: [provBind.reasonCode!],
      discriminatorFingerprint: null,
    };
  }

  const expectedFp = rule4D3D5DiscriminatorFingerprintV1Hash(envelope);
  if (envelope.discriminatorFingerprint !== expectedFp) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: ['D3_D5_DISCRIMINATOR_FINGERPRINT_INVALID'],
      discriminatorFingerprint: null,
    };
  }

  const sens = envelope.sensitivityAssessmentStatus;
  if (
    sens === 'MISSING' ||
    sens === 'NOT_EVALUATED' ||
    sens === 'INVALID' ||
    sens === 'AMBIGUOUS' ||
    sens === 'CONTRADICTORY'
  ) {
    const code =
      sens === 'MISSING' || sens === 'NOT_EVALUATED'
        ? 'D3_D5_SENSITIVITY_MISSING'
        : 'D3_D5_SENSITIVITY_AMBIGUOUS';
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: [code],
      discriminatorFingerprint: expectedFp,
    };
  }

  if (
    envelope.closeD05DiscriminatorStatus === 'CONTRADICTORY' ||
    envelope.closeD05DiscriminatorStatus === 'UNRESOLVED' ||
    !sensitivityCloseAligned(envelope)
  ) {
    return {
      status: 'unresolved',
      dilution: null,
      reasonCodes: ['D3_D5_DISCRIMINATOR_CONTRADICTORY'],
      discriminatorFingerprint: expectedFp,
    };
  }

  if (sens === 'ASSESSED_D5_QUALIFIED' && envelope.closeD05DiscriminatorStatus === 'QUALIFIES_D5') {
    reasons.push('CLOSE_D05_QUALIFIES_D5');
    return {
      status: 'resolved',
      dilution: 'D5',
      reasonCodes: reasons,
      discriminatorFingerprint: expectedFp,
    };
  }

  if (
    sens === 'ASSESSED_D5_NOT_QUALIFIED' &&
    envelope.closeD05DiscriminatorStatus === 'QUALIFIES_D3'
  ) {
    reasons.push('CLOSE_D05_D3_PATH');
    return {
      status: 'resolved',
      dilution: 'D3',
      reasonCodes: reasons,
      discriminatorFingerprint: expectedFp,
    };
  }

  return {
    status: 'unresolved',
    dilution: null,
    reasonCodes: ['D3_D5_DISCRIMINATOR_UNRESOLVED'],
    discriminatorFingerprint: expectedFp,
  };
}
